mpa/cls/trainer.py

```diff
@@ -230,6 +235,9 @@ def train_worker(gpu, dataset, cfg, distributed, validate, timestamp, meta):
         # an ugly walkaround to make the .log and .log.json filenames the same
         runner.timestamp = f'{timestamp}'
 
+        # remember the max number of train images. It is needed for EMA hook heuristic
```

Comment:
Unfortunately, there is no such attribute in "before run" stage. So, there is two options: 1) add some not pretty lines to trainer for counting max number of images and do not create EMA model if the specified criteria is not valid 2) Create EMA model always and check the criteria in the `after_train_iter` stage and also we need to modify some how `after_train_epoch` in `eval_hook`. 

I perefer 1st option. But you can propose another solution

Reply:
Hmm, here is some proposal and modification for yours
1) If another hook or algorithm should be aware about the number of images, current code can be nice for us. But as far as I know, there is no hook or algorithm that should know number of images before training so current code is being only for EMA hook. The fact "only for EMA hook" is not problem for me but we can choose another method like 2), without some modification in trainer.py

2) It's more reasonable for me, if we can update EMA hook instead of EvalHook. My imagine is here :
In EMA hook,
` before_train_epoch(self, runner):`
` if not hasattr(runner.number_of_images):`
` runner.number_of_images = len(runner.data_loader.dataset)`
I believe we can be able to avoid checking the criteria in every epoch or iteration.

Reply:
I got rid of any changes in trainer.py and transfer it to EMA hook `before_train_epoch` method. Also, I got rid of creating EMA model in case of criteria of the dataset length are not fulfilled